CT Surface Sectioned Geometry

larswik opened the pull request:
Proposal for new CT "Surface Sectioned Geometry". Addresses #224
In this process I also propose a new RepresentationIdentifier = "SectionedSurface". A little bit "unbalanced" vs the corresponding for sectioned solids which are defined using "AdvancedSweptSolid", a level missing for surfaces. Since Surface Sectioned Geometry was proposed to be placed as a sibling to "Surface3D geometry"/RepresentationIdentifier="Surface3D" and "Surface Tesselation Geometry" /RepresentationIdentifier="Tesselation", I figured that this would be my proposal.

larswik committed:
CT Surface Sectioned Geometry

Proposal for new CT "Surface Sectioned Geometry". Addresses #224
In this process I also propose a new RepresentationIdentifier = "SectionedSurface". A little bit "unbalanced" vs the corresponding for sectioned solids which are defined using "AdvancedSweptSolid", a level missing for surfaces. Since Surface Sectioned Geometry was proposed to be placed as a sibling to "Surface3D geometry"/RepresentationIdentifier="Surface3D" and "Surface Tesselation Geometry" /RepresentationIdentifier="Tesselation", I figured that this would be my proposal.

jmirtsch reviewed (changes requested):
Fix UniqueIds

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml

Co-authored-by: Jon Mirtschin <jonm@geometrygym.com>

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml

Co-authored-by: Jon Mirtschin <jonm@geometrygym.com>

SergejMuhic commented:
The proposed changes require a schema change and documentation change also.

larswik commented:
Besides the documentation change, is it IfcShapeRepresentationTypes that you are referring to? I felt that the way things are structured for RepresentationType, it felt to me as if something new was needed (maybe even ideally an "AdvancedSweptSurface"). However if it is ok to just use "SurfaceModel", then fine by me. I am not entirely sure how these are actually used in software. Happy to remove that part of the proposal.

SergejMuhic commented:
Yes, that is what I am referring to. Adding 'SectionedSurface' means we would modify the function.

SergejMuhic commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,77 @@
+																<DocModelRuleEntity Name="IfcProfileDef" />
+															</Rules>
+														</DocModelRuleAttribute>
+														<DocModelRuleAttribute Name="CrossSectionPositions" Identification="CrossSectionPositions">
```
This I would remove. Same as Sectioned solid. It is restricted in the template already.

AlexBrad1eyCT replied:
have been removed and resolved

SergejMuhic commented:
> 
> 
> Besides the documentation change, is it IfcShapeRepresentationTypes that you are referring to? I felt that the way things are structured for RepresentationType, it felt to me as if something new was needed (maybe even ideally an "AdvancedSweptSurface"). However if it is ok to just use "SurfaceModel", then fine by me. I am not entirely sure how these are actually used in software. Happy to remove that part of the proposal.

The purpose of the identifier and type has somewhat diminished with mvdxml. Not sure if we are ready to remove it completely (especially now) since it does provide a mechanism to quite quickly identify whether the implementation supports it. This kind of also implies the necessity of "SectionedSurface" as type since `IfcSectionedSurface` is not your typical surface representation. Let me know what you want to do.

larswik committed:
Update DocTemplateDefinition.xml

Remove CrossSectionPositions

larswik commented:
Removed CrossSectionPositions. Remains the issue about RepresentationType.

larswik committed:
IfcShapeRepresentationTypes (SectionedSurface Rep type) & IfcShapeRepresentation indentation

Added SectionedSurface to IfcShapeRepresentationTypes (please check).
Indentations of Curve2D, Curve3D, Surface2D, Surface3D in the RepresentationType table
Added SectionedSurface under Surface as a new RepresentationType.

AlexBrad1eyCT commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,72 @@
+										</DocModelRuleAttribute>
+										<DocModelRuleAttribute CardinalityMin="1" CardinalityMax="1" Name="RepresentationIdentifier" Identification="Identifier">
+											<Rules>
+												<DocModelRuleEntity Name="IfcLabel">
+													<Rules>
+														<DocModelRuleAttribute Name="Value=Surface" Description="Value=Surface" />
+													</Rules>
+												</DocModelRuleEntity>
```
This is the wrong encoding it should look the same as the RepresentationType Below, this needs correcting

larswik replied:
Seems like there is an error in the encoding at the level "Surface Geometry", right? Didn't touch that but will look into it.

AlexBrad1eyCT replied:
yes if you merge the develop you should get the correct reference on Surface Geometry that just needs adding like in sectionedsolidHorizontal

AlexBrad1eyCT replied:
```suggestion
												<DocModelRuleEntity Name="IfcLabel">
													<Rules>
														<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>
													</Rules>
												</DocModelRuleEntity>
```

AlexBrad1eyCT reviewed (changes requested):
Only the error on the inherited RepresentationIdentifier attribute needs correcting.

larswik committed:
Merge remote-tracking branch 'upstream/develop' into CT-Surface-Sectioned-Geometry

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml

Co-authored-by: Alex Bradley <alex.bradley@abcon.tech>

larswik commented:
Out of my comfort zone here but this still looks wrong (Surface Geometry):
![image](https://user-images.githubusercontent.com/25506404/140072939-d957a79f-d76e-4cc0-9632-3ec727838302.png)
Shouldn't it be more in the format (Body Geometry):
![image](https://user-images.githubusercontent.com/25506404/140073054-be78b6bc-88e9-494e-b114-d2a5ff2f7008.png)
However the cml encoding looks very similar...

AlexBrad1eyCT reviewed (approved):
LFTM

AlexBrad1eyCT commented:
I saw that and compared the XML and its exactly the same (minus the ids and the literal ofcourse) so i cannot explain it.
some thing for the backlog i think @jmirtsch @SergejMuhic

larswik commented:
I've made the change and the xml encoding looks ok. However, still missing the unique id generation so haven't pushed yet. I am in the Development branch of IfcDoc. Thought that was the correct one.

AlexBrad1eyCT commented:
@larswik the xml is all correct the uniqueId attribute is only on the original rule so that is fine its just the display in IfcDoc seems to have a periodic error.

AlexBrad1eyCT commented:
@larswik you must not be running the latest build, I just opened and saved your branch for this pull request and it did not change the sectioned surface rule. double check you have the latest development of IfcDoc and that it has been rebuilt the fix was only merged (PR87) on monday.

larswik commented:
@AlexBrad1eyCT could you try to change the RepresentationIdentifier part of the Sectioned surface rule? It does not seem right at the moment. I have fetched and pulled everything from IfcDoc/Development.

AlexBrad1eyCT commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:
```diff
@@ -0,0 +1,72 @@
+										</DocModelRuleAttribute>
+										<DocModelRuleAttribute CardinalityMin="1" CardinalityMax="1" Name="RepresentationIdentifier" Identification="Identifier">
+											<Rules>
+												<DocModelRuleEntity Name="IfcLabel">
+													<Rules>
+														<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>
+													</Rules>
+												</DocModelRuleEntity>
```
Found the Issue Double bracketing should be fine now @larswik
```suggestion
												<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>>
```

AlexBrad1eyCT replied:
![image](https://user-images.githubusercontent.com/74732183/140729042-9df6e814-f556-4de7-8d81-c0631180268d.png)

AlexBrad1eyCT commented:
@larswik please review and submit my fix.
@jmirtsch @SergejMuhic @pjanck please review and approve where necessary ASAP so we can get this PR closed and be able deliver the CTs

larswik committed:
Merge remote-tracking branch 'upstream/develop' into CT-Surface-Sectioned-Geometry

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml

Co-authored-by: Alex Bradley <alex.bradley@abcon.tech>

larswik commented:
Thanks @AlexBrad1eyCT ! 
This LFTM now.

AlexBrad1eyCT reviewed (approved):
LFTM please review everyone!

jmirtsch reviewed (commented):
It looks fine to me.

pjanck commented on IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/Documentation.md:
```diff
@@ -0,0 +1,7 @@
+The _Surface Sectioned Geometry_ is the representation for elements having a having a 'Surface' representation where the surface is specified using two or more open profiles that are swept between specified positions along the directrix, e.g. for defining the upper boundary of an _IfcCourse_.
```
```suggestion
The _Surface Sectioned Geometry_ is the representation for elements having a 'Surface' representation, where the surface is specified using two or more open profiles that are swept between specified positions along the directrix, e.g. for defining the upper boundary of an _IfcCourse_.
```

pjanck reviewed (changes requested):
Small typo in the documentation, otherwise LFTM

larswik committed:
Update IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/Documentation.md

Co-authored-by: Stefan Jaud <59165496+pjanck@users.noreply.github.com>
